# Googleplay data analysis

This is Goolgeplay data analysis

- Create a dictionary where each key is a unique app name and the corresponding dictionary value is the highest number of reviews of that app.

- Start by creating an empty dictionary named `reviews_max`.
Loop through the Google Play data set (make sure you don't include the header row). For each iteration:
Assign the app name to a variable named name.
Convert the number of reviews to float. Assign it to a variable named `n_reviews`.

If name already exists as a key in the reviews_max dictionary and reviews_max[name] < n_reviews, update the number of reviews for that entry in the reviews_max dictionary.
If name is not in the reviews_max dictionary as a key, create a new entry in the dictionary where the key is the app name, and the value is the number of reviews. Make sure you don't use an else clause here, otherwise the number of reviews will be incorrectly updated whenever reviews_max[name] < n_reviews evaluates to False.
Inspect the dictionary to make sure everything went as expected. Measure the length of the dictionary — remember that the expected length is 9,659 entries.
Use the dictionary you created above to remove the duplicate rows:

Start by creating two empty lists: android_clean (which will store our new cleaned data set) and already_added (which will just store app names).
Loop through the Google Play data set (make sure you don't include the header row), and for each iteration:
Assign the app name to a variable named name.
Convert the number of reviews to float, and assign it to a variable named n_reviews.
If n_reviews is the same as the number of maximum reviews of the app name (the number can be found in the reviews_max dictionary) and name is not already in the list already_added (read the solution notebook to find out why we need this supplementary condition):
Append the entire row to the android_clean list (which will eventually be a list of list and store our cleaned data set).
Append the name of the app name to the already_added list — this helps us to keep track of apps that we already added.

In [1]:
from csv import reader

Applefile = open('AppleStore.csv', encoding="utf8")
readApple = reader(Applefile)
apple_data = list(readApple)

Googlefile = open('googleplaystore.csv', encoding="utf8")
readGoogle = reader(Googlefile)
google_data = list(readGoogle)
#google_data = [google_data[0:10474], google_data[10474:]]

def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))
        
explore_data(apple_data, 1, 3)
explore_data(google_data, 1, 3)

n_row_apple = len(apple_data)
n_column_apple = len(apple_data[0])

n_row_google = len(google_data)
n_column_google = len(google_data[0])

print(n_row_apple, n_column_apple, n_row_google, n_column_google, '\n')

explore_data(apple_data, 0, 1)
explore_data(google_data, 0, 1)

#del apple_data[100:]
#del google_data[100:]

# n_row_apple = len(apple_data)
# n_column_apple = len(apple_data[0])

# n_row_google = len(google_data)
# n_column_google = len(google_data[0])

# print(n_row_apple, n_column_apple, n_row_google, n_column_google, '\n')


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


7198 16 10842 13 

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Ge

In [2]:
duplicate_apps = []
unique_apps = []
for app in google_data[1:]:
    name = app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)

print(duplicate_apps[0:2])

['Quick PDF Scanner + OCR FREE', 'Box']


In [3]:
n_duplicate = len(duplicate_apps)
print(len(duplicate_apps))

1181


In [4]:
print('Expected length:', len(google_data[1:]) - n_duplicate)

Expected length: 9660


In [5]:
reviews_max = {}
for app in google_data[1:]:
    name = app[0]
    n_reviews = app[7]
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
    if name not in reviews_max:
        reviews_max[name] = n_reviews
        
print(len(reviews_max))
    

9660


In [6]:
android_clean = []
already_added = []
for app in google_data[1:]:
    name = app[0]
    n_reviews = app[7]
    if n_reviews == reviews_max[name] and name not in already_added:
        android_clean.append(app)
        already_added.append(name)
print(len(android_clean))

9660


In [7]:
print(android_clean[0])

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


In [8]:
def check_character(a_string):
    non_ch = 0
    for character in a_string:
        if ord(character) > 127:
            non_ch += 1
    if non_ch > 3:
        return False
    else:
        return True

print(check_character('Instagram'))
print(check_character('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(check_character('Docs To Go™ Free Office Suite'))
print(check_character('Instachat 😜'))

True
False
True
True


In [9]:
g_en_apps = []
for app in google_data[1:]:
    name = app[1]
    if check_character(name):
        g_en_apps.append(app)
        
a_en_apps = []
for app in apple_data[1:]:
    name = app[1]
    if check_character(name):
        a_en_apps.append(app) 
        
print(len(g_en_apps), len(a_en_apps))

10841 6183


In [10]:
def removedollar(a_string):
#    print(a_string)
    if ord(a_string[0]) == ord('$'):
#        print(a_string[1:])
        return a_string[1:]
    if ord(a_string[0]) == ord('E'):
#        print('0')
        return '0'
#    print(a_string)
    return a_string

In [11]:
g_free_apps = []
for app in google_data[1:]:
    price = app[7]
    price = removedollar(price)
    price = float(price)
    if price == 0.0:
        g_free_apps.append(app)
        
a_free_apps = []
for app in apple_data[1:]:
    price = float(app[7])
    if price == 0.0:
        a_free_apps.append(app) 
        
print(len(g_free_apps), len(a_free_apps))

10041 929


In [12]:
google_genre = {}
for key in google_data[1:]:
    genre = key[11]
    if genre in google_genre:
        google_genre[genre] += 1
    else:
        google_genre[genre] = 1

apple_genre = {}
for key in apple_data[1:]:
    genre = key[11]
    if genre in apple_genre:
        apple_genre[genre] += 1
    else:
        apple_genre[genre] = 1
        
print(len(google_genre))
print(len(apple_genre))

2834
23


In [13]:
def freq_table(dataset, index):
    data_freq = {}
    for key in dataset[1:]:
        name = key[index]
        if name in data_freq:
            data_freq[name] += 1
        else:
            data_freq[name] = 1
    return data_freq

def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

#print(google_data[0][9])
display_table(apple_data, 11)
print('\n')
display_table(google_data, 1)
print('\n')
display_table(google_data, 9)

Games : 3862
Entertainment : 535
Education : 453
Photo & Video : 349
Utilities : 248
Health & Fitness : 180
Productivity : 178
Social Networking : 167
Lifestyle : 144
Music : 138
Shopping : 122
Sports : 114
Book : 112
Finance : 104
Travel : 81
News : 75
Weather : 72
Reference : 64
Food & Drink : 63
Business : 57
Navigation : 46
Medical : 23
Catalogs : 10


FAMILY : 1972
GAME : 1144
TOOLS : 843
MEDICAL : 463
BUSINESS : 460
PRODUCTIVITY : 424
PERSONALIZATION : 392
COMMUNICATION : 387
SPORTS : 384
LIFESTYLE : 382
FINANCE : 366
HEALTH_AND_FITNESS : 341
PHOTOGRAPHY : 335
SOCIAL : 295
NEWS_AND_MAGAZINES : 283
SHOPPING : 260
TRAVEL_AND_LOCAL : 258
DATING : 234
BOOKS_AND_REFERENCE : 231
VIDEO_PLAYERS : 175
EDUCATION : 156
ENTERTAINMENT : 149
MAPS_AND_NAVIGATION : 137
FOOD_AND_DRINK : 127
HOUSE_AND_HOME : 88
LIBRARIES_AND_DEMO : 85
AUTO_AND_VEHICLES : 85
WEATHER : 82
ART_AND_DESIGN : 65
EVENTS : 64
PARENTING : 60
COMICS : 60
BEAUTY : 53
1.9 : 1


Tools : 842
Entertainment : 623
Education : 549


# Analyze the frequency table you generated for the prime_genre column of the App Store data set.

- What is the most common genre? What is the runner-up?

Games : 3862

- What other patterns do you see?

The harder the genre, the less the frequency

- What is the general impression — are most of the apps designed for practical purposes (education, shopping, utilities, productivity, lifestyle) or more for entertainment (games, photo and video, social networking, sports, music)?

People still pay attention to education, ...

- Can you recommend an app profile for the App Store market based on this frequency table alone? If there's a large number of apps for a particular genre, does that also imply that apps of that genre generally have a large number of users?

Skip

In [14]:
apple_ft = freq_table(apple_data, 11)
for genre in apple_ft:
    total = 0
    len_genre = 0
    for key in apple_data[1:]:
        genre_app = key[11]
        if genre_app == genre:
            total += float(key[7])
            len_genre += 1
    avg_rating = total / len_genre
    print(genre, avg_rating)



Lifestyle 2.8055555555555554
Photo & Video 3.8008595988538683
Games 3.6850077679958573
Book 2.4776785714285716
Weather 3.5972222222222223
Entertainment 3.2467289719626167
Utilities 3.278225806451613
Navigation 2.6847826086956523
Reference 3.453125
Medical 3.369565217391304
Shopping 3.540983606557377
Social Networking 2.9850299401197606
News 2.98
Business 3.745614035087719
Sports 2.982456140350877
Health & Fitness 3.7
Food & Drink 3.1825396825396823
Productivity 4.00561797752809
Catalogs 2.1
Music 3.9782608695652173
Travel 3.376543209876543
Finance 2.4326923076923075
Education 3.376379690949227


In [23]:
#print(google_data[10473][5])
google_Category_ft = freq_table(google_data, 1)
for category in google_Category_ft:
    total = 0
    len_category = 0
    for key in google_data[1:10474]:
        category_app = key[1]
        if category_app == category:
            n_installs = key[5]
            n_installs = n_installs.replace('+', '')
            n_installs = n_installs.replace(',', '')
            if n_installs[0] == 'F':
                n_installs = '0'
            n_installs = float(n_installs)
#            print(n_installs)
            total += n_installs
            print(total)
            len_category += 1
    print(total, 'he ', len_category)
#    avg_installs = total / len_category
#    print(category, avg_installs)
        

10000000.0
60000000.0
70000000.0
75000000.0
76000000.0
126000000.0
131000000.0
231000000.0
232000000.0
233000000.0
233500000.0
243500000.0
248500000.0
249500000.0
299500000.0
309500000.0
310500000.0
320500000.0
325500000.0
330500000.0
335500000.0
340500000.0
340600000.0
341600000.0
342600000.0
343600000.0
353600000.0
358600000.0
359600000.0
364600000.0
364700000.0
374700000.0
375700000.0
376700000.0
376800000.0
381800000.0
382800000.0
382900000.0
432900000.0
437900000.0
438000000.0
439000000.0
444000000.0
445000000.0
455000000.0
465000000.0
466000000.0
466100000.0
466200000.0
476200000.0
476250000.0
481250000.0
486250000.0
491250000.0
491350000.0
491450000.0
492450000.0
492950000.0
493050000.0
493550000.0
494050000.0
495050000.0
496050000.0
497050000.0
497150000.0
497250000.0
498250000.0
499250000.0
509250000.0
519250000.0
519350000.0
519450000.0
524450000.0
524500000.0
525500000.0
526500000.0
527000000.0
537000000.0
547000000.0
547050000.0
552050000.0
557050000.0
562050000.0
562150000